### PR TITLE
[SR-5688] [Sema] Handle key path component base type on MemberAccessOnOptionalBaseFailure

### PR DESCRIPTION
**Pull request by LucianoPAlmeida**
The fix was being recorded but it was failing to handle the key path component type.
This just handles the `KeyPathComponent` locator.

Resolves SR-5688.

### CODE REVIEW

File: lib/Sema/CSDiagnostics.cpp
Comment (hamishknight) [inline lib/Sema/CSDiagnostics.cpp:1010]: We want to avoid querying the constraint system during diagnostics, as its state may not be the same as when the fix was originally recorded. Note also that this doesn't handle the case where the optional base is the root of the key path, e.g `\String?.count`.

I think the easiest solution here might be to have the `UnwrapOptionalBase` fix store the optional base type rather than trying to recover it later. Note though that we may still need some custom logic to get the right source location for where to insert the postfix `?` and `!` fix-its in the key path (and to disable the fix-its for an optional key path root). Could you adjust the tests to make sure we're inserting them in the right place?

**Review by hamishknight — COMMENTED** [review]
Thanks for the fix!

File: lib/Sema/CSDiagnostics.cpp
Comment (LucianoPAlmeida) [reply lib/Sema/CSDiagnostics.cpp:1010]: > Note though that we may still need some custom logic to get the right source location for where to insert the postfix ? and ! fix-its in the key path (and to disable the fix-its for an optional key path root). Could you adjust the tests to make sure we're inserting them in the right place?

Sure, I'm adjusting that right now :) 

> We want to avoid querying the constraint system during diagnostics, as its state may not be the same as when the fix was originally recorded. Note also that this doesn't handle the case where the optional base is the root of the key path, e.g \String?.count.

Right, I'll make the adjustments :) Thank you

Comment (LucianoPAlmeida) [reply lib/Sema/CSDiagnostics.cpp:1010]: Also @hamishknight 

> Note also that this doesn't handle the case where the optional base is the root of the key path, e.g \String?.count

I can't see the case where that will require this fix, do you have an example?

Comment (hamishknight) [reply lib/Sema/CSDiagnostics.cpp:1012]: @LucianoPAlmeida Does `\String?.count` not create the fix? As I understand it, we should add a member constraint for `count` with a base of `String?`, requiring the member-on-optional fix.

File: lib/Sema/CSDiagnostics.cpp
Comment (hamishknight) [inline lib/Sema/CSDiagnostics.cpp:1010]: Unfortunately I don't think this will get the correct location for a subscript component in e.g:

```swift
struct S {
  subscript(_ x: Int) -> String? { "" }
}

\S.[5].count
```

I think we should add a `getEndLoc` method to `KeyPathExpr::Component` to handle this, which can get the subscript's end loc from its index expr, what do you think?

File: lib/Sema/CSDiagnostics.cpp
Comment (LucianoPAlmeida) [reply lib/Sema/CSDiagnostics.cpp:1012]: Ahh sorry, I just misunderstood ... because I was thinking of the fixes and how in this situation it will help to add an extra `?` or `!` ...
But yes it creates it

Comment (hamishknight) [reply lib/Sema/CSDiagnostics.cpp:1012]: Right yeah, we'll want to emit the error about the incorrect optionality, but we'll want to avoid telling the user to add `!` or `?` :)

File: lib/Sema/CSDiagnostics.cpp
Comment (LucianoPAlmeida) [reply lib/Sema/CSDiagnostics.cpp:1010]: Humm, I haven't thought in that case

File: lib/Sema/CSDiagnostics.cpp
Comment (LucianoPAlmeida) [reply lib/Sema/CSDiagnostics.cpp:1012]: That's what I thought :) Thanks @hamishknight, I'll make some quick adjustments in another PR and get back to this

**Comment by LucianoPAlmeida** [comment]
@hamishknight @xedin There are still some tests failing I have to look, so I'll ping you when this is ready for a review :)

**Comment by LucianoPAlmeida** [comment]
Right, this is ready :) @hamishknight  @xedin

File: include/swift/AST/DiagnosticsSema.def
Comment (theblixguy) [inline include/swift/AST/DiagnosticsSema.def:1075]:
```suggestion
    "remove optional '?' from key path root base type %0", (Type))
```

File: lib/Sema/CSDiagnostics.cpp
Comment (theblixguy) [inline lib/Sema/CSDiagnostics.cpp:989]: Nit: You can use `getLocator()` directly here.

File: lib/Sema/CSDiagnostics.cpp
Comment (theblixguy) [inline lib/Sema/CSDiagnostics.cpp:1047]:
```suggestion
      auto unwrappedName = nominalDecl->getBaseName();
```

File: include/swift/AST/DiagnosticsSema.def
Comment (theblixguy) [reply include/swift/AST/DiagnosticsSema.def:1075]: Also, “change Foo? to Foo” might be a simpler message.

Comment (LucianoPAlmeida) [reply include/swift/AST/DiagnosticsSema.def]: Nice catch, thanks!

> Also, “change key path base type from Foo? to Foo” maybe?

 I think this is the implicit message already provided by the fixItReplace we have "String?" -> "String" so I'm not sure it would add much adding this in the note. And since what we want to indicate is remove optionality, I think the note should say something along these lines :)

File: lib/Sema/CSDiagnostics.cpp
Comment (hamishknight) [inline lib/Sema/CSDiagnostics.cpp:997]: I think it would be better if this was an override of `FailureDiagnostic::getSourceRange`.

File: lib/Sema/CSDiagnostics.cpp
Comment (hamishknight) [inline lib/Sema/CSDiagnostics.cpp:1005]: I believe you'll want to use `getRootType` instead of `getParsedRoot` to correctly handle nested type roots, e.g:

```swift
struct S {
  struct Nested {
    var foo = ""
  }
}

\S.Nested?.foo.count
```

File: lib/Sema/CSDiagnostics.cpp
Comment (hamishknight) [inline lib/Sema/CSDiagnostics.cpp:1044]: What if it's not a nominal root, such as `\(Int, Int)?.0`? Rather than digging out a decl, we should be able to print the unwrapped type directly through `unwrappedBaseType.getString()`.

File: lib/Sema/CSDiagnostics.cpp
Comment (hamishknight) [inline lib/Sema/CSDiagnostics.cpp:1059]: Note that these diagnostics will use the source location provided by `FailureDiagnostic::getLoc`, which will use the entire key path's location. I think it would make sense to override `getLoc` for `MemberAccessOnOptionalBaseFailure` and have it return the end loc of the source range we compute.

File: lib/Sema/CSDiagnostics.cpp
Comment (hamishknight) [reply lib/Sema/CSDiagnostics.cpp:1005]: Ah we'll also need to handle the case where the explicit root type is null to handle cases like this:

```swift
func takesKP(_ kp: KeyPath<String?, Int>) {}
takesKP(\.count)
```

File: lib/Sema/CSDiagnostics.cpp
Comment (LucianoPAlmeida) [reply lib/Sema/CSDiagnostics.cpp]: Fixed! :)

**Comment by LucianoPAlmeida** [comment]
@hamishknight Thank you for the review 
Sorry for taking too long to answer, but I've addressed all things. 
So this is ready for review again :)
cc @xedin

**Comment by LucianoPAlmeida** [comment]
Humm, @xedin @hamishknight there is still one test case failing I have to check

**Comment by LucianoPAlmeida** [comment]
Right, now tests are happy so this is ready for review :)

File: lib/Sema/CSDiagnostics.cpp
Comment (hamishknight) [inline lib/Sema/CSDiagnostics.cpp:1042]: I'm not sure this suggestion is correct, AFAIK we don't support optional chaining on the root of a key path.

File: lib/Sema/CSDiagnostics.h
Comment (hamishknight) [inline lib/Sema/CSDiagnostics.h:504]: Do we need to special case key path components here? IMO it would make more sense if we returned `getSourceRange().End` for both cases here as then we'll always point to the dot in the member access.

File: lib/Sema/CSSimplify.cpp
Comment (hamishknight) [inline lib/Sema/CSSimplify.cpp:6953]: What case is this handling? I'm not sure we should be silently unwrapping the base without recording a fix for it.

**Review by hamishknight — COMMENTED** [review]
Looks great! Just a couple more comments

File: lib/Sema/CSDiagnostics.h
Comment (LucianoPAlmeida) [reply lib/Sema/CSDiagnostics.h:504]: Nice, I just had this way because I thought that for non key path components the range was the member range, so it would point to the end of the member no the dot at the end of the base.

File: lib/Sema/CSDiagnostics.cpp
Comment (LucianoPAlmeida) [reply lib/Sema/CSDiagnostics.cpp:1042]: Ah true, so I think we can't really suggest any fix here ... :( 
I would think about inserting the unwrapped type here but I just check it gives 
```swift
func SR5688_KP(_ kp: KeyPath<String?, Int>) {}

SR5688_KP(\String.count) // Key path value type 'String?' cannot be converted to contextual type 'String'
```
Which I don't think is correct, or it is at the very least a bad diagnostic ... not sure

File: lib/Sema/CSSimplify.cpp
Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp:6953]: But as for my understanding, we shouldn't attempt fixes when type matching is `TMF_ApplyingFix`  ... this handles the case 

```swift
class T {
    func m1() {
      let l = self.m2!.prop1
    }

    func m2() -> U! {
      return U()
    }
}
```
Where if `m2` is a function type with no argument we attempt to simplify member with the result type `U?` as  base to verify if it has a `prop1` member in order to suggest if this could have a "make it a call" fix `self.m2()!.prop1` in this case. In this case, we only want to match the types and not record fixes(that is why I think the `TMF_ApplyingFix` is passed) which will end up in the wrong diagnostic, so that is why I think this is fine because it will still record fixes in normal no `TMF_ApplyingFix` case.

File: lib/Sema/CSDiagnostics.cpp
Comment (LucianoPAlmeida) [reply lib/Sema/CSDiagnostics.cpp]: Just remove it, not really sure if we could suggest something else here ... 
cc @xedin

Comment (LucianoPAlmeida) [reply lib/Sema/CSDiagnostics.cpp]: @hamishknight :)
I was thinking more about this today and to me, maybe misunderstanding the diagnostics, but this still feels like a problem. I'm not really sure if that is something that supposed to work or if it is a matter of improving diagnostics for this case. What do you think? Should we open a ticket to track this if it is indeed a problem?

File: lib/Sema/CSDiagnostics.h
Comment (LucianoPAlmeida) [reply lib/Sema/CSDiagnostics.h]: Ah, I've adjusted that, seems like it is ok :)

File: lib/Sema/CSDiagnostics.cpp
Comment (hamishknight) [reply lib/Sema/CSDiagnostics.cpp]: Yeah, I agree it could be better – perhaps we should have a custom diagnostic that emphasises that the problem is with the key path's root type being inferred to an optional type? Ideally we'd be able to point to where the inferred type is coming from in the user's code, but we don't yet track that information in the constraint system.

File: lib/Sema/CSDiagnostics.h
Comment (hamishknight) [reply lib/Sema/CSDiagnostics.h]: Great!

File: lib/Sema/CSSimplify.cpp
Comment (hamishknight) [reply lib/Sema/CSSimplify.cpp:6953]: @LucianoPAlmeida Ah I see! 

I don't think using `TMF_ApplyingFix` here is the right fix, it's only really used when additional fixes are unlikely to be useful (and in such cases it returns `SolutionKind::Error` if a fix would be required). IMO in an ideal world we wouldn't need `TMF_ApplyingFix` at all, as we should be able to rely on ranking fixes by their impact.

For the example you give, it seems to me that it might be worth attempting to record an `InsertExplicitCall` fix within `simplifyOptionalObjectConstraint` if we're trying to perform an optional unwrap on a function type of the form `() -> T?`, which would allow us to re-write the code as `self.m2()!.prop1` without telling the user add or remove a `!`.

File: lib/Sema/CSDiagnostics.cpp
Comment (LucianoPAlmeida) [reply lib/Sema/CSDiagnostics.cpp]: > Yeah, I agree it could be better – perhaps we should have a custom diagnostic that emphasises that the problem is with the key path's root type being inferred to an optional type?

I think we could do that, as you mentioned the ideal would be to point out the declaration and suggest not use optional as keypath root. I'll adjust the message asap then :)

File: lib/Sema/CSSimplify.cpp
Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp:6953]: > For the example you give, it seems to me that it might be worth attempting to record an InsertExplicitCall fix within simplifyOptionalObjectConstraint if we're trying to perform an optional unwrap on a function type of the form () -> T?, which would allow us to re-write the code as self.m2()!.prop1 without telling the user add or remove a !.


Right thank you @hamishknight, I'll have to think about this ... TBH I still quite didn't fully understand. Because even doing this we still have to verify if the return of `()-> U?` in this case `U?` has a `prop1` member in order to know if make it a call `()` is applicable, which still would be trying to `simplifyMemberConstraint` with `U?` as a base and would lead to wrong diagnostics when failing on any of the fixed constraints... I think I just have to think about this a bit more, thanks @hamishknight

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp]: @hamishknight Since the only thing we are doing here is to verify if `U` from result type `U?` has a member, I think we could do is strip the optional out for the check because this shouldn't matter in this situation and we avoid producing the misleading diagnostic. I just updated the last commit, let me know what you think :)

Comment (hamishknight) [reply lib/Sema/CSSimplify.cpp]: > Because even doing this we still have to verify if the return of `()-> U?` in this case `U?` has a `prop1` member in order to know if make it a call `()` is applicable, which still would be trying to `simplifyMemberConstraint` with `U?` as a base and would lead to wrong diagnostics when failing on any of the fixed constraints

You should be able to unwrap the result type of the function when applying the `InsertExplicitCall` fix in `simplifyOptionalObjectConstraint` and match it against the type `second`, which will allow it to propagate to the member constraint which should then succeed with the fix in place.

File: lib/Sema/CSSimplify.cpp
Comment (hamishknight) [inline lib/Sema/CSSimplify.cpp:6974]: I'm not sure silently stripping optionality here is the right solution, as it could cause us to miss recording useful optional unwrapping fixes.

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp:6974]: Ah, you are right here it is wrong, but that is not where it meant to be ... I'll adjust it :)

File: lib/Sema/CSSimplify.cpp
Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp]: Right, I can take a look on this :)
But this would mean we have to remove the logic from `simplifyMemberConstraint` that currently diagnose this?

Comment (hamishknight) [reply lib/Sema/CSSimplify.cpp]: @LucianoPAlmeida This would be in addition to the logic in `simplifyMemberConstraint`. Trying the fix in `simplifyOptionalObjectConstraint` would handle general cases where a function call needs to be inserted before the unwrap e.g `x()!` or `x()?.y`, whereas the logic in `simplifyMemberConstraint` handles cases where a function call needs to be inserted before the member access such as `x!().y`.

We could try to reduce the duplicated logic by factoring it out into an `InsertExplicitCall::attempt` method (which could take a parameter determining whether the function result needs unwrapping).

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp]: Ahh I see! Thanks @hamishknight, I think I got it now. Will send a patch ASAP

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp]: > You should be able to unwrap the result type of the function when applying the `InsertExplicitCall` fix in `simplifyOptionalObjectConstraint` and match the unwrapped type against the type `second`, which will allow it to propagate to the member constraint which should then succeed with the fix in place.

What do you mean by match? I'm just trying to wrap my head around how the member constraint would be propagated here.

Comment (hamishknight) [reply lib/Sema/CSSimplify.cpp]: @LucianoPAlmeida You'll want to call `matchTypes` and check if the result is a success before recording the fix, similar to how it is done in `repairFailures` (though you'll want a `ConstraintKind::Bind` instead of `ConstraintKind::Conversion`:

https://github.com/apple/swift/blob/c825d7987eba360c77e8343e106addaadad4491e/lib/Sema/CSSimplify.cpp#L3141-L3146

The propagation of the base type should then happen automatically.

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp]: Hum, that is what I thought, in fact, that is what I'm doing, so I think I'm not totally lost as I thought on this one. Thank you @hamishknight :)

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp]: @hamishknight Adjusted 👍

File: lib/Sema/CSDiagnostics.cpp
Comment (LucianoPAlmeida) [reply lib/Sema/CSDiagnostics.cpp]: @hamishknight I just added a TODO diagnostics for this, it is ok or do you think it is something that should be handled here on this patch?

File: lib/Sema/CSSimplify.cpp
Comment (hamishknight) [reply lib/Sema/CSSimplify.cpp]: Thanks! I have a bit more feedback, but otherwise this is looking great :)

File: lib/Sema/CSDiagnostics.cpp
Comment (hamishknight) [reply lib/Sema/CSDiagnostics.cpp]: @LucianoPAlmeida I'm totally fine with leaving it as a TODO for now – would you mind also filing it as a bug/improvement?

Comment (LucianoPAlmeida) [reply lib/Sema/CSDiagnostics.cpp]: Yeap, Done https://bugs.swift.org/browse/SR-13079 :)

**Review by hamishknight — APPROVED** [review]
Thanks again! I'll let @xedin have the final say, but this looks good to me!

Please make sure to do a squash-and-merge when merging to ensure the git history is kept clean (congrats on commit access btw!).

**Comment by LucianoPAlmeida** [comment]
> Please make sure to do a squash-and-merge when merging to ensure the git history is kept clean (congrats on commit access btw!).

Sure, Thank you for the review @hamishknight :) I've learned some nice new things here \o/

File: lib/Sema/CSDiagnostics.h
Comment (xedin) [inline lib/Sema/CSDiagnostics.h:497]: There is `resolveType` which does exactly this :) I think it would make sense to do that during construction of MemberAccessOnOptionalBaseFailure like so `MemberBaseType(resolveType(memberBaseType))`, that way this method could simply return `MemberBaseType` instead of having to do the same over multiple times...

**Review by xedin — APPROVED** [review]
I have left a minor comment otherwise LGTM!

File: lib/Sema/CSDiagnostics.h
Comment (LucianoPAlmeida) [reply lib/Sema/CSDiagnostics.h:497]: Great! I'll adjust, Thank you @xedin :)

Comment (LucianoPAlmeida) [reply lib/Sema/CSDiagnostics.h]: Fixed @xedin :)

**Comment by LucianoPAlmeida** [comment]
@swift-ci please smoke test

File: lib/Sema/CSDiagnostics.h
Comment (xedin) [reply lib/Sema/CSDiagnostics.h]: Great!

**Comment by LucianoPAlmeida** [comment]
@xedin The windows build didn't trigger or it just didn't show on the github UI like that one in the other PR?

**Comment by xedin** [comment]
Looks like it just didn't report the status - https://ci-external.swift.org/view/Pull%20Request/job/swift-PR-windows/4504/

**Comment by xedin** [comment]
@swift-ci please smoke test Windows platform

**Comment by LucianoPAlmeida** [comment]
Squashed and merged, Thank you @hamishknight @xedin :)

**Pull request merged**
